File: 프로그래머스/2/92342. 양궁대회/양궁대회.java
class Solution {
    
    static final int[] IMPOSSIBLE = {-1};
    static int[] ans = IMPOSSIBLE;
    static int maxDiff = 0;
    
    public int[] solution(int n, int[] info) {
        
        solve(0, new int[11], n, info);
        return ans;
    }
    
    public void solve(int index, int[] ryan, int n, int[] apeach) {

        if (index == ryan.length) {
            
            //남은 화살이 있을 때
            if (n > 0) return;

            //라이언과 어피치의 점수 차
            int diff = getDiff(ryan, apeach);
            
            if (diff <= 0 || diff < maxDiff) {
                return;
            }

            if (diff > maxDiff || check(ryan)) {
                maxDiff = diff;
                
                if (ans == IMPOSSIBLE) ans = new int[11];

                System.arraycopy(ryan, 0, ans, 0, ryan.length);
                // ans = ryan; 주의! 얕은 복사 X
            }
            
            return;
        }

        for (int i = 0; i <= n; i++) {
            ryan[index] = i;  //각 점수에 0개 ~ N개를 맞힐 수 있다.
            solve(index + 1, ryan, n - i, apeach);
            ryan[index] = 0;  //원상복구
        }
    }
    
    public boolean check(int[] ryan) {

        for (int i = 10; i >= 0; i--) {
            if (ans[i] == ryan[i]) continue;
            return ryan[i] > ans[i];
        }

        return false;
    }

    public int getDiff(int[] ryan, int[] apeach) {

        int diff = 0;

        for (int i = 0; i < apeach.length; i++) {

            if (ryan[i] == 0 && apeach[i] == 0) continue;

            if (ryan[i] <= apeach[i]) {
                diff -= (10 - i);
            } else {
                diff += (10 - i);
            }
        }

        return diff;
    }
}

File: 프로그래머스/2/92342. 양궁대회/양궁대회.py
from copy import deepcopy

IMPOSSIBLE = [-1]
ans = IMPOSSIBLE
maxDiff = 0

def get_diff(ryan, apeach):

    diff = 0

    for i in range(len(apeach)):
        if ryan[i] == 0 and apeach[i] == 0:
            continue

        if ryan[i] <= apeach[i]:
            diff -= (10 - i)
        else:
            diff += (10 - i)

    return diff


def check(ryan):

    for i in range(10, -1, -1):
        if ans[i] == ryan[i]:
            continue

        return ryan[i] > ans[i]

    return False


def solve(index, ryan, n, apeach):

    if index == len(ryan):
        if n > 0:
            return
        
        global maxDiff, ans

        diff = get_diff(ryan, apeach)

        if diff <= 0 or diff < maxDiff:
            return

        if diff > maxDiff or check(ryan):            
            maxDiff = diff
            ans = deepcopy(ryan)
        
        return


    for i in range(0, n + 1):
        ryan[index] = i
        solve(index + 1, ryan, n - i, apeach)
        ryan[index] = 0

def solution(n, info):

    solve(0, [0] * 11, n, info)
    return ans
